client/db/src/storage_cache/key_next.rs

```diff
@@ -0,0 +1,865 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2019-2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Structure and implementation used for caching `next_key` calls.
+//! This uses ordered mapping with key intervals, see `CachedInterval`.
+
+
+use std::collections::{HashMap, BTreeMap};
+use sp_core::storage::ChildInfo;
+use super::{EstimateSize,
+};
+
+pub(super) struct LRUOrderedKeys<K> {
+	/// We use a BTreeMap for storage internally.
+	intervals: BTreeMap<K, Box<KeyOrderedEntry<K>>>,
+	/// Intervals for child storages.
+	child_intervals: HashMap<Vec<u8>, BTreeMap<K, Box<KeyOrderedEntry<K>>>>,
+	/// Current total size of contents.
+	used_size: usize,
+	/// Limit size of contents.
+	limit: usize,
+	/// Dummy `KeyOrderedEntry` containing `next` pointer
+	/// as the oldest entry.
+	/// `prev` pointer is used as the lru entry, meaning
+	/// if `prev` equals to `next` the lru structure is empty.
+	lru_bound: Box<KeyOrderedEntry<K>>,
+}
+
+#[derive(Default)]
+pub(super) struct LocalOrderedKeys<K: Ord> {
+	/// We use a BTreeMap for storage internally.
+	intervals: BTreeMap<K, CachedInterval>,
+	/// Intervals for child storages.
+	child_intervals: HashMap<Vec<u8>, BTreeMap<K, CachedInterval>>,
+}
+	
+struct KeyOrderedEntry<K> {
+	/// Entry accessed before.
+	prev: *mut KeyOrderedEntry<K>,
+	/// Entry access after.
+	next: *mut KeyOrderedEntry<K>,
```

Comment:
I think using an unsafe linked list is an overkill here.  Using `Option<K>` would probably result in a much simpler code which is fast enough. Even though iteration will require btree lookups, it will still be an order of magnitude faster than trie lookups.

Reply:
Using Option<K> or Option<Rc<K> would surely work, but I am not sure how code will be simpler, IIRC there is two primitives: `detach` when removing and `touch` for accessing or adding. Without unsafe, api may be a bit more verbose for mutable access reason, but generally the same.
 In https://github.com/paritytech/substrate/compare/master...cheme:next_key_single_cache, I started merging cache size limit to avoid having size ratio.
 In this branch all cache structs uses the same lru linked list, so we do not have 3 different cache size to define (I realize when doing it that the ratio for child was not (or no longer) exposed as a parameter).
 The branch uses the same implementation, but similarily we can replace lru pointer by some key reference (lru implementation is pretty similar to the one of linkedhashmap)?